File: app/src/main/java/com/coinninja/coinkeeper/model/dto/CompletedBroadcastDTO.java
package com.coinninja.coinkeeper.model.dto;

import android.os.Parcel;
import android.os.Parcelable;

import com.coinninja.bindings.TransactionData;
import com.coinninja.coinkeeper.service.client.model.Contact;

import java.util.Objects;

public class CompletedBroadcastDTO extends BroadcastTransactionDTO implements Parcelable {

    public static final Creator<CompletedBroadcastDTO> CREATOR = new Creator<CompletedBroadcastDTO>() {
        @Override
        public CompletedBroadcastDTO createFromParcel(Parcel in) {
            return new CompletedBroadcastDTO(in);
        }

        @Override
        public CompletedBroadcastDTO[] newArray(int size) {
            return new CompletedBroadcastDTO[size];
        }
    };
    public String transactionId;

    public CompletedBroadcastDTO(BroadcastTransactionDTO broadcastActivityDTO, String transactionId) {
        super(broadcastActivityDTO.getTransactionData(), broadcastActivityDTO.getContact(),
                broadcastActivityDTO.isMemoShared(), broadcastActivityDTO.getMemo(),
                broadcastActivityDTO.getPublicKey());
        this.transactionId = transactionId;
    }

    public CompletedBroadcastDTO(TransactionData transactionData, String transactionId, Contact contact) {
        super(transactionData, contact);
        this.transactionId = transactionId;
    }

    protected CompletedBroadcastDTO(Parcel in) {
        super(in);
        transactionId = in.readString();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeString(transactionId);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), transactionId);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        CompletedBroadcastDTO that = (CompletedBroadcastDTO) o;
        return Objects.equals(transactionId, that.transactionId);
    }

    public boolean shouldShareMemo() {
        return getContact() != null && getContact() != null && !"".equals(getMemo()) && isMemoShared()
                && !"".equals(getPublicKey());
    }

    public boolean hasMemo() {
        return getMemo() != null && !"".equals(getMemo());
    }

    public String getPhoneNumberHash() {
        return getContact().getHash();
    }

    public boolean hasPublicKey() {
        return !"".equals(getPublicKey());
    }
}

File: app/src/main/java/com/coinninja/coinkeeper/model/dto/CompletedBroadcastDTO.kt
package com.coinninja.coinkeeper.model.dto

import android.os.Parcelable
import app.coinninja.cn.libbitcoin.model.TransactionData
import app.dropbit.annotations.Mockable
import com.coinninja.coinkeeper.model.Identity
import kotlinx.android.parcel.Parcelize

@Mockable
@Parcelize
data class CompletedBroadcastDTO(
        var transactionData: TransactionData,
        var isMemoShared: Boolean = false,
        var memo: String? = "",
        var identity: Identity? = null,
        var publicKey: String? = null,
        var transactionId: String? = null) : Parcelable {


    constructor(broadcastActivityDTO: BroadcastTransactionDTO, transactionId: String) : this(
            broadcastActivityDTO.transactionData,
            broadcastActivityDTO.isMemoShared,
            broadcastActivityDTO.memo,
            broadcastActivityDTO.identity,
            broadcastActivityDTO.publicKey,
            transactionId)

    constructor(transactionData: TransactionData, transactionId: String, identity: Identity) : this(
            transactionData, false, "", identity, null, transactionId
    )


    fun shouldShareMemo(): Boolean {
        return (identity != null && !memo.isNullOrEmpty() && isMemoShared && !publicKey.isNullOrEmpty())
    }

    fun hasMemo(): Boolean {
        return !memo.isNullOrEmpty()
    }

    fun hasPublicKey(): Boolean {
        return !publicKey.isNullOrEmpty()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as CompletedBroadcastDTO

        if (transactionData != other.transactionData) return false
        if (isMemoShared != other.isMemoShared) return false
        if (memo != other.memo) return false
        if (identity != other.identity) return false
        if (publicKey != other.publicKey) return false
        if (transactionId != other.transactionId) return false

        return true
    }

    override fun hashCode(): Int {
        var result = transactionData.hashCode()
        result = 31 * result + isMemoShared.hashCode()
        result = 31 * result + (memo?.hashCode() ?: 0)
        result = 31 * result + (identity?.hashCode() ?: 0)
        result = 31 * result + (publicKey?.hashCode() ?: 0)
        result = 31 * result + (transactionId?.hashCode() ?: 0)
        return result
    }
}
